File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/com/example/factory/documents/PdfDocument.java
package com.example.factory.documents;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
public class PdfDocument implements Document {
    private LocalDateTime creationTime;
    private String fileName;
    public PdfDocument() {
        this.creationTime = LocalDateTime.now();
        this.fileName = "Document_" + creationTime.format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")) + ".pdf";
    }
    @Override
    public void open() {
        System.out.println("Opening PDF document: " + fileName);
        System.out.println("Loading PDF reader application...");
    }
    @Override
    public void save() {
        System.out.println("Saving PDF document: " + fileName);
        System.out.println("Document saved in .pdf format");
    }
    @Override
    public void close() {
        System.out.println("Closing PDF document: " + fileName);
        System.out.println("PDF reader application closed");
    }
    @Override
    public void print() {
        System.out.println("Printing PDF document: " + fileName);
        System.out.println("High-quality print mode enabled...");
    }
    @Override
    public String getDocumentType() {
        return "Portable Document Format";
    }
    @Override
    public LocalDateTime getCreationTime() {
        return creationTime;
    }
}

File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/Documents/PdfDocument.cs
using System;
using System.Globalization;

namespace FactoryDemo.Documents;

public sealed class PdfDocument : IDocument
{
    private readonly DateTime _creationTime = DateTime.Now;
    private readonly string _fileName;

    public PdfDocument()
    {
        _fileName = $"Document_{_creationTime.ToString("yyyyMMdd_HHmmss", CultureInfo.InvariantCulture)}.pdf";
    }

    public void Open()  => Console.WriteLine($"Opening PDF document: {_fileName}\nLoading PDF reader application…");
    public void Save()  => Console.WriteLine($"Saving PDF document: {_fileName}\nDocument saved in .pdf format");
    public void Close() => Console.WriteLine($"Closing PDF document: {_fileName}\nPDF reader application closed");
    public void Print() => Console.WriteLine($"Printing PDF document: {_fileName}\nHigh‑quality print mode enabled…");

    public string GetDocumentType()  => "Portable Document Format";
    public DateTime GetCreationTime() => _creationTime;
}
